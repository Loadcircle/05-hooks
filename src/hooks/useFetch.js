import { useEffect, useRef, useState } from "react"

export const useFetch = (url, initialState = {data: null, loading: true, error: null}) => {

    const isMounted = useRef(true)

    const [state, setState] = useState(initialState);

    useEffect(()=>{

        return()=>{
            isMounted.current = false;
        }

    }, [])

    useEffect(() => {
        setState({...initialState, loading: true})
        fetch(url)
            .then(resp=>resp.json())
            .then(data=>{
                if(isMounted.current){
                    setState({
                        loading: false,
                        error: null,
                        data,
                    });
                }
            })
    }, [url])

    return state;
    
}
