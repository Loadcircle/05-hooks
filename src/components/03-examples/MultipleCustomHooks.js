import React from 'react' 
import { useFetch } from '../../hooks/useFetch';
import { useCounter } from '../../hooks/useCounter';

import './../components.css';


export const MultipleCustomHooks = () => {
    const url = 'https://www.breakingbadapi.com/api/quotes/';

    const {state:page, increment:nextPage} = useCounter(1);

    const {loading, data} = useFetch(url+page);

    const {author:autor, quote:cita} = !!data && data[0];
      
    return (
        <div>
            <h1>Custom Hooks!!!</h1>
            
            {
                loading 
                ?
                (
                    <div className="alert alert-info text-center">
                        Loading...
                    </div>
                )
                : 
                (
                    <blockquote className="blockquote text-right">
                        <p className="mb-0">
                            {cita}
                        </p>
                        <footer className="blockquote-footer">{autor}</footer>
                    </blockquote>    
                )
            }

            <button className="btn btn-primary" onClick={()=>{nextPage(1)}}>Siguiente cita</button>

        </div>
    )
}
