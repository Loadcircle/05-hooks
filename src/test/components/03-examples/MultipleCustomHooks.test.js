import { shallow } from 'enzyme';
import React from 'react';
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';
jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');

describe('Pruebas en MultipleCustomHooks', ()=>{

    useFetch.mockReturnValue({
        data: null,
        loading: true, 
        error: null
    });
    useCounter.mockReturnValue({
        state: 10,
        increment: ()=>{}
    });

    const wrapper = shallow(<MultipleCustomHooks/>)

    test('Debe de mostrarse correctamente', ()=>{

        expect(wrapper).toMatchSnapshot();

    });

    test('Debe de mostrar la info', ()=>{

        useFetch.mockReturnValue({
            data: [{
                author: 'Jesus',
                quote: 'Hola Mundo'
            }],
            loading: false, 
            error: null
        });

        const wrapper = shallow(<MultipleCustomHooks/>);

        expect(wrapper.find('.alert').exists()).toBe(false);
        expect(wrapper.find('.mb-0').text().trim()).toBe('Hola Mundo');
        expect(wrapper.find('.blockquote-footer').text().trim()).toBe('Jesus');   

    });

});