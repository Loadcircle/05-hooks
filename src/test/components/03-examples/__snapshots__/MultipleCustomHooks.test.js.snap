// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en MultipleCustomHooks Debe de mostrarse correctamente 1`] = `
<div>
  <h1>
    Custom Hooks!!!
  </h1>
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente cita
  </button>
</div>
`;
